lib/wasi/src/os/task/process.rs

```diff
@@ -0,0 +1,372 @@
+    syscalls::platform_clock_time_get,
+    WasiThread, WasiThreadHandle, WasiThreadId,
+};
+
+use super::{control_plane::ControlPlaneError, task_join_handle::TaskJoinHandle};
+
+/// Represents the ID of a sub-process
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct WasiProcessId(u32);
+
+impl WasiProcessId {
+    pub fn raw(&self) -> u32 {
+        self.0
+    }
+}
+
+impl From<i32> for WasiProcessId {
+    fn from(id: i32) -> Self {
+        Self(id as u32)
+    }
+}
+
+impl Into<i32> for WasiProcessId {
+    fn into(self) -> i32 {
+        self.0 as i32
+    }
+}
+
+impl From<u32> for WasiProcessId {
+    fn from(id: u32) -> Self {
+        Self(id)
+    }
+}
+
+impl Into<u32> for WasiProcessId {
+    fn into(self) -> u32 {
+        self.0 as u32
+    }
+}
+
+impl std::fmt::Display for WasiProcessId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+/// Represents a process running within the compute state
+// TODO: fields should be private and only accessed via methods.
+#[derive(Debug, Clone)]
+pub struct WasiProcess {
+    /// Unique ID of this process
+    pub(crate) pid: WasiProcessId,
+    /// ID of the parent process
+    pub(crate) ppid: WasiProcessId,
+    /// The inner protected region of the process
+    pub(crate) inner: Arc<RwLock<WasiProcessInner>>,
+    /// Reference back to the compute engine
+    // TODO: remove this reference, access should happen via separate state instead
+    // (we don't want cyclical references)
+    pub(crate) compute: WasiControlPlane,
+    /// Reference to the exit code for the main thread
+    pub(crate) finished: Arc<TaskJoinHandle>,
+    /// List of all the children spawned from this thread
+    pub(crate) children: Arc<RwLock<Vec<WasiProcessId>>>,
+    /// Number of threads waiting for children to exit
+    pub(crate) waiting: Arc<AtomicU32>,
+}
+
+// TODO: fields should be private and only accessed via methods.
+#[derive(Debug)]
+pub struct WasiProcessInner {
+    /// The threads that make up this process
+    pub threads: HashMap<WasiThreadId, WasiThread>,
+    /// Number of threads running for this process
+    pub thread_count: u32,
+    /// Seed used to generate thread ID's
+    pub thread_seed: WasiThreadId,
+    /// All the thread local variables
+    pub thread_local: HashMap<(WasiThreadId, TlKey), TlVal>,
+    /// User data associated with thread local data
+    pub thread_local_user_data: HashMap<TlKey, TlUser>,
+    /// Seed used to generate thread local keys
+    pub thread_local_seed: TlKey,
+    /// Signals that will be triggered at specific intervals
+    pub signal_intervals: HashMap<Signal, WasiSignalInterval>,
+    /// Represents all the process spun up as a bus process
+    pub bus_processes: HashMap<WasiProcessId, Box<BusSpawnedProcess>>,
+    /// Indicates if the bus process can be reused
+    pub bus_process_reuse: HashMap<Cow<'static, str>, WasiProcessId>,
```

Comment:
Why are all this fields public, in an inner type? It seems wrong

Reply:
I think I already added a comment above that these should be private. 

( I very recently refactored the `WasiThread` in the same vain - reduce redundant separate Arcs, make fields private)

Reply:
Refactor tracked in #3607